modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/ReadOnlyTransactionImpl.java

```diff
@@ -113,22 +113,31 @@ public TablePartitionId commitPartition() {
     @Override
     public CompletableFuture<Void> commitAsync() {
         return TransactionsExceptionMapperUtil.convertToPublicFuture(
-                finish(true, readTimestamp, false),
+                finish(true, readTimestamp, false, false),
                 TX_COMMIT_ERR
         );
     }
 
     @Override
     public CompletableFuture<Void> rollbackAsync() {
         return TransactionsExceptionMapperUtil.convertToPublicFuture(
-                finish(false, readTimestamp, false),
+                finish(false, readTimestamp, false, false),
                 TX_ROLLBACK_ERR
         );
     }
 
     @Override
-    public CompletableFuture<Void> finish(boolean commit, HybridTimestamp executionTimestamp, boolean full) {
+    public CompletableFuture<Void> rollbackTimeoutExceededAsync() {
+        return TransactionsExceptionMapperUtil.convertToPublicFuture(
+                finish(false, readTimestamp, false, true),
+                TX_ROLLBACK_ERR
+        ).thenAccept(unused -> this.timeoutExceeded = true);
```

Comment:
Why does this assignment happen after the transaction finish completes?

Reply:
Do you think I should do it before calling 'finish'.

Reply:
Well, I think first it should be decided whether 'before' and 'after' are equivalent :) What triggered me here is that it seems that 'actually finishing and marking as `timeoutExceeded`' do not look atomic. What happens if someone sees that the transaction is already finished, but does not see the effect of the flag assignment yet? Are other races possible?

Reply:
This is a part of the internal API and we use this flag as a "cached" value. So, I believe this must be set to `true` after the actual rollback happened. 

Speaking about race here, I can't say there are any problem with it here. If you have some ideas/tests scenarios I would appreciate if you share them.

Reply:
Rewrote that part. Moved the flag change into `finish` method for Read Tx and into `finishInternal` for Read Write Tx.